File: 0155-min-stack/0155-min-stack.rb
class MinStack
    def initialize()
        @stack = []
        @min = []
    end


=begin
    :type val: Integer
    :rtype: Void
=end
    def push(val)
        @stack << val
        
        if @min.length == 0 || val < @min[-1]
            @min << val
        else
            @min << @min[-1]
        end
    end


=begin
    :rtype: Void
=end
    def pop()
        @min.pop
        @stack.pop
    end


=begin
    :rtype: Integer
=end
    def top()
        @stack[-1]
    end


=begin
    :rtype: Integer
=end
    def get_min()
        @min[-1]
    end


end

# Your MinStack object will be instantiated and called as such:
# obj = MinStack.new()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.get_min()

File: 0155-min-stack/0155-min-stack.js
// like a regular stack, but also keeps track of running min
// elements in the MinStack are [current min, value]

// update the min as needed when pushing a new value in
// or when pushing the first value, also use it as initial min

// the other operations can use standard array operations
// but return either the min or actual value

var MinStack = function() {
    this.stack = [];
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    if (!this.stack.length) {
        this.stack.push([val, val]);
    } else { 
        const prevMin = this.stack[this.stack.length - 1][0];
        this.stack.push([Math.min(prevMin, val), val]);
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    return this.stack.pop()[1];
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.stack[this.stack.length-1][1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.stack[this.stack.length-1][0];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */